### ISSUE TITLE
MTP doesn't work on Samsung devices

### ISSUE THREAD

**Issue by phhusson** [phhusson/treble_experimentations#1121]
Android's Mtp server has:

```
    bool ffs_ok = access(FFS_MTP_EP0, W_OK) == 0;
    if (ffs_ok) {
        bool aio_compat = android::base::GetBoolProperty("sys.usb.ffs.aio_compat", false);
        mHandle = aio_compat ? new MtpFfsCompatHandle(controlFd) : new MtpFfsHandle(controlFd);
    } else {
        mHandle = new MtpDevHandle();
    }
```
Possibly, forcing one codepath or the another would fix the issue.

**Comment by Avrution** [comment]
Guess this explains why I could not get my device to show as a drive in my system.

**Comment by famouzkk** [comment]
anything about this? it can be related to logcat error "MtpServer: Failed to start usb driver!"? i installed various GSI's and MTP dont work; in own kernel tried samsung mtp drivers and google mtp drivers but still nothing, in recovery only google drivers works, any workaround?

**Comment by phhusson** [comment]
Doing the equivalent of this should work: (though I tried and failed for the moment)
https://github.com/TeamWin/android_device_samsung_beyond0lte/commit/bf23185c0ae811ad021efc19988d48bd0531017d

**Comment by famouzkk** [comment]
I tried that too but still fails.

**Comment by phhusson** [comment]
what have you tried precisely?

**Comment by famouzkk** [comment]
Swaping samsung drivers with google ones and applying this patch to init.recovery.usb.rc, should i try with default ones(samsung) and this ffs patch?(i mean i tried this but i only booted in recovery, and in recovery mtp was not recognizable, i didint had time and flashed back kernel with google drivers)

**Comment by phhusson** [comment]
Well if you apply it to init.recovery.usb.rc, it's expected not to work...
Apply it on the user .usb.rc

**Comment by englishlukeh** [comment]
Is there any solution to this yet? Will patching user.usb.rc fix MTP for Samsung devices?

**Comment by ponces** [comment]
Changed all lines of init.usb.configfs.rc that mentions:
`/config/usb_gadget/g1/functions/mtp.gs0`
to
`/config/usb_gadget/g1/functions/ffs.mtp`

Had no luck :(

**Comment by ponces** [comment]
> Changed all lines of init.usb.configfs.rc that mentions:
> `/config/usb_gadget/g1/functions/mtp.gs0`
> to
> `/config/usb_gadget/g1/functions/ffs.mtp`
> 
> Had no luck :(
> 
> I have no root so I'm not able to check which functions are available at /config/usb_gadget/g1/functions.
> Can anybody check this for me please?

Nevermind. Fixed MTP by using non-Samsung MTP drivers in kernel.

**Comment by phhusson** [comment]
It's more complicated than changing mtp.gs0 to ffs.mtp if you look at the twrp patch.

**Comment by ponces** [comment]
Reverted back to stock boot to test again.
I already tried by adding:

```
on boot
    mkdir /config/usb_gadget/g1/functions/ffs.mtp 0770 shell shell
    mkdir /dev/usb-ffs/mtp 0770 mtp mtp
    mount functionfs mtp /dev/usb-ffs/mtp uid=2000,gid=2000
```
or
```
on boot
    mkdir /config/usb_gadget/g1/functions/ffs.mtp
    mkdir /dev/usb-ffs/mtp 0770 mtp mtp
    mount functionfs mtp /dev/usb-ffs/mtp rmode=0770,fmode=0660,uid=1024,gid=1024,no_disconnect=1
```

No success but it appears to be close as Windows identified a device named "MTP" although no device appears in the explorer...

**Comment by szmarczak** [comment]
I'm experiencing the same on A40. Even USB debugging does not work. However, sometimes it keeps reconnecting. If I pick `USB tethering`, it BSODs Windows every time. And not just Windows, Fedora freezes and is unresponsive. Although most of the time it just does not work and nothing happens.

**Comment by szmarczak** [comment]
Oooh this time it did not reconnect! `lsusb` shows 

```
Bus 001 Device 120: ID 04e8:6860 Samsung Electronics Co., Ltd Galaxy series, misc. (MTP mode)
```

however

![image](https://github.com/user-attachments/assets/cfaf8d33-fe35-4404-b118-b3e1a9c44959)

**Comment by szmarczak** [comment]
![image](https://github.com/user-attachments/assets/9e67856d-e974-43f2-a086-8641acb8480e)

ADB works.

**Comment by szmarczak** [comment]
![image](https://github.com/user-attachments/assets/888658ee-c62e-4243-a6cd-33c5465aa00b)

Even scrcpy works.

**Comment by szmarczak** [comment]
dmesg after I changed the USB mode to PTP and it stopped working:
```
[  847.334083] init: processing action (sys.usb.config=adb && sys.usb.configfs=1 && sys.usb.ffs.ready=1) from (/system/etc/init/hw/init.usb.configfs.rc:20)
[  847.337865] usb: clear_current_usb_link: GSI_image : Clear cfg->func_list 
[  847.337890] usb: clear_current_usb_link : f->name = ptp
[  847.337901] usb: make_adb_connection_for_gsi: ADB Only mode for recovery or VTS test
[  847.337915] usb: get_adb_function_from_linked_func : f_name = uts
[  847.337926] usb: get_adb_function_from_linked_func : f_name = rndis
[  847.337937] usb: get_adb_function_from_linked_func : f_name = dm
[  847.337947] usb: get_adb_function_from_linked_func : f_name = conn_gadget
[  847.337957] usb: get_adb_function_from_linked_func : f_name = midi
[  847.337967] usb: get_adb_function_from_linked_func : f_name = accessory
[  847.337977] usb: get_adb_function_from_linked_func : f_name = audio_source
[  847.337987] usb: get_adb_function_from_linked_func : f_name = ncm
[  847.337997] usb: get_adb_function_from_linked_func : f_name = mtp
[  847.338007] usb: get_adb_function_from_linked_func : f_name = acm
[  847.338016] usb: get_adb_function_from_linked_func : f_name = ptp
[  847.338026] usb: config_usb_cfg_link adb function setting fails -17
[  847.338238] init: Command 'symlink /config/usb_gadget/g1/functions/ffs.adb /config/usb_gadget/g1/configs/b.1/f1' action=sys.usb.config=adb && sys.usb.configfs=1 && sys.usb.ffs.ready=1 (/system/etc/init/hw/init.usb.configfs.rc:22) took 2ms and failed: symlink() failed: File exists
[  847.339467] gadget_dev_desc_UDC_store: +++
[  847.339519] Config b/1 of g1 needs at least one function.
[  847.339548] configfs-gadget 13600000.dwc3: failed to start g1: -22
[  847.339558] gadget_dev_desc_UDC_store: err return ---
[  847.339882] init: Command 'write /config/usb_gadget/g1/UDC ${sys.usb.controller}' action=sys.usb.config=adb && sys.usb.configfs=1 && sys.usb.ffs.ready=1 (/system/etc/init/hw/init.usb.configfs.rc:23) took 1ms and failed: Unable to write to file '/config/usb_gadget/g1/UDC': Unable to write file contents: Invalid argument
```

tl;dr: `'/config/usb_gadget/g1/UDC': Unable to write file contents: Invalid argument`

[dmesg.txt](https://github.com/user-attachments/files/16503181/dmesg.txt)

The dmesg also contains logs when adb was working.

**Comment by szmarczak** [comment]
Interestingly, after a reboot ADB is working again. I have no idea what's going on lol

**Comment by szmarczak** [comment]
USB tethering is selected but there's no network device on Windows nor Linux :(

**Comment by szmarczak** [comment]
If I'm on Windows, if I choose MTP it switchs back to `No data transfer`. But it stays on Linux, however when I try to open the device in file explorer it says `Unable to open MTP device`. If I choose `USB tethering`, it switchs back to `No data transfer` on Linux too.

**Comment by szmarczak** [comment]
With https://github.com/kevios12/android_kernel_samsung_universal7885/releases it works... I wonder what secret sauce did they use to make this work 🤔
